Cloudinary-pwa-kit/packages/pwa-cloudinary/app/components/cloudinary-image-gallery/index.jsx

```diff
@@ -1,37 +1,53 @@
-import React, { useState } from 'react'
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import React, {useState, useEffect} from 'react'
 import PropTypes from 'prop-types'
-import {
-    AspectRatio,
-    Box,
-    Img,
-    Flex,
-
-    // Hooks
-    ListItem,
-    List,
-    useMultiStyleConfig
-} from '@chakra-ui/react'
+import {AspectRatio, Box, Img, Flex, ListItem, List, useMultiStyleConfig} from '@chakra-ui/react'
 import DynamicImage from '@salesforce/retail-react-app/app/components/dynamic-image'
-
 import RenderCloudinaryGalleryWidget from '../../components/cloudinary-widgets'
 import RenderCloudinaryVideoPlayer from '../../components/cloudinary-product-video'
+import {cloudinary} from '../../../config/default'
 
 const EnterKeyNumber = 13
 
 /**
  * The image gallery displays The Image Gallery Coming from Cloudinary in Product-Detail Page.
  */
-const CloudinaryImageGallery = ({ size, cloudinaryImageGallery = {} }) => {
+const CloudinaryImageGallery = ({size, cloudinaryImageGallery = {}}) => {
     const [selectedIndex, setSelectedIndex] = useState(0)
+    const [productLoaded, setProductLoaded] = useState(false)
     let imageUrls
-    const styles = useMultiStyleConfig('ImageGallery', { size })
+
+    useEffect(() => {
+        if (
+            cloudinaryImageGallery?.galleryEnabled &&
+            document.querySelector(
+                `script[src="${`https://product-gallery.cloudinary.com/${cloudinary.versions.CLDGalleryVersion}/all.js`}"]`
+            ) === null
+        ) {
+            const script = document.createElement('script')
+            script.src = `https://product-gallery.cloudinary.com/${cloudinary.versions.CLDGalleryVersion}/all.js`
+            script.async = true
+            document.head.appendChild(script)
+        }
+
+        setTimeout(() => {
+            setProductLoaded(true)
+        }, 700)
```

Comment:
what is the purpose of this delay? are we suspected to race conditions?

Reply:
this is due to react states, the product data loads first and our library defers loading, we need to have a small delay to load the CLD library and then product data.

Reply:
so this should happen after the product data loads using an event listener or async/await pattern.
just setting 700ms seems risky that the app would be either slow when the product loads fast, or broken when the product loads slow

Reply:
this delay is for product load

Reply:
and it will only trigger once the product loads

Reply:
the product gallery loading should not uploaded with script.async, and should not uploaded with a delay. We can consider uploading it with script.deferred in the future.
We have in line 22: const [productLoaded, setProductLoaded] = useState(false)
doesn't that wait for the state to end?

This whole thing seems fragile

Reply:
The scenario is, we have added a gallery-enabled check for PGW CDN, and the gallery-enabled bit comes with product data, once the product data comes to the page we check if the gallery is enabled or not to load the CDN and in the meantime the product data loads on the screen, in that case, CDN takes a few ms to load in DOM in the gallery is enabled, and the gallery HTML renders before that, we just a little delay to load the CDN and then gallery widget code.

Reply:
we can include PGW CDN on PDP without gallery enabled check, and it will work fine

Reply:
@yuval-cloudinary we have added a change, please have a look if it looks good to you

Reply:
@asad-rafter I think that I now understand better the 700ms delay, and I think that we may still encounter the situation that the component loads the legacy gallery when it wait for the Cloudinary gallery to load the JS.
The solution would be using a 2nd useEffect statement that depends on the first one.

And there is a 2nd useEffect at Cloudinary-pwa-kit/packages/pwa-cloudinary/app/components/cloudinary-widgets/index.jsx but this is done only after the condition ([l50](https://github.com/cloudinary/cloudinary_sfcc_pwa-kit/blob/bdaeb01e164afc60dd8cbc324b7ca5ef333561d1/Cloudinary-pwa-kit/packages/pwa-cloudinary/app/components/cloudinary-image-gallery/index.jsx#L50): cloudinaryImageGallery.galleryEnabled && productLoaded ) was already executed.

What do you think?

Reply:
I believe it would work, because it will only call the Cloudinary-pwa-kit/packages/pwa-cloudinary/app/components/cloudinary-widgets/index.jsx component when the Gallery Script loaded. I have also rename the state for better understanding and meaningful code.

Reply:
OK, so please deploy it to the MRT and we will test it